### ISSUE TITLE
Volume Control Scales Unnaturally

### ISSUE THREAD

**Issue by aphirst** [exaile/exaile#709]
### Steps to Reproduce (for bugs)

1. Use Exaile with Pulseaudio (the Exaile in-program volume control affects the volume of the source/sink that's assigned to Exaile by Pulseaudio directly)
2. Start a familiar music track playing, with reasonably consistent volume throughout
3. Lower Exaile's volume to 0 via the in-application dial
4. Raise Exaile's volume by the smallest possible notch (scroll wheel, one pixel, whatever)

### Expected Behavior

The playback volume should start very low (both in the stated value in a Pulseaudio volume control application and in terms of how it actually sounds), and increase in proportion to how far you increase the dial.

### Current Behavior

With only one "notch" increased in Exaile, the Pulseaudio volume level is already 37% (and it sounds pretty loud, of course). With a second notch it increases to 37%, then 46%, then 53% and it's diminishing returns from there.

### Possible Solution

Sound volume can be measured in a few different ways, linearly and logarithmically, to reflect how we actually perceive the sound. It seems as if this is being overcorrected by Exaile somewhere in the volume control code. The % of Exaile's volume meter should correspond to the % of Pulseaudio's volume source/sink because Pulseaudio is **already** correcting for the difference between decibels and amplitude/energy/whatever.

### Environment

* Operating System and version: Arch Linux (x86_64, rolling release)

* Exaile Version: 4.1.0-alpha1

**Comment by sjohannes** [comment]
As far as I know, we're passing the volume number to GStreamer without doing any logarithmic scaling, but I'm not very familiar with the engine code. @virtuald, any idea?

**Comment by virtuald** [comment]
I don't believe we're doing anything special. @aphirst maybe try disabling fade in the playback settings? That's the only place I'm aware of that anything is multiplied with volume.

There's a commented out debug statement in [engine.py](https://github.com/exaile/exaile/blob/ccde3c0f018d9d61dd47e9f04c9f818b908c8fd3/xl/player/gst/engine.py#L546), might try uncommenting it and see if the internal volume matches the pulse volume?

**Comment by aphirst** [comment]
@virtuald I don't have fade enabled.
![ex](https://user-images.githubusercontent.com/1735556/83610569-8dfbc480-a577-11ea-8b12-40c5572a486b.png)

I think you meant L546:
```
self.logger.debug("Set playbin volume: %.2f", volume)
```
But it didn't seem to affect anything? I ran exaile in the console and expected some kind of output as I changed the volume but it doesn't seem to appear anywhere.
```

**Comment by unkie** [comment]
I had the same issue with the volume scaling and was able to fix this behaviour.
After debugging I found that when exaile sets the volume on the GST PlayBin property, it is using the Cubic Volume Format.
This is what causes the "5% in exaile is 37% in pavucontrol".
By setting the volume explicitly as Cubic, we get linear behaviour. Very odd...
I'm guessing this is a bug in the python bindings of gstreamer.

Patch: https://github.com/unkie/exaile/commit/998c304f3ba723406374157a3fe98c4defc744e6

**Comment by unkie** [comment]
I did some more digging, and it looks like a bug in `gstreamer` to me.

We have two options to set the volume:

1. Set the volume property of the PlayBin: [`self.playbin.props.volume = real`](https://github.com/unkie/exaile/commit/998c304f3ba723406374157a3fe98c4defc744e6#diff-ce3af07e86acd3a0b14ac9b414e603ddL549)
2. Use the method [`set_volume()`](https://github.com/unkie/exaile/commit/998c304f3ba723406374157a3fe98c4defc744e6#diff-ce3af07e86acd3a0b14ac9b414e603ddR554)

Method 2 will convert the volume scaling, and then do the equivalent of method 1: set the volume property.

When the property is set, [`gst_pulsesink_set_volume()`](https://gitlab.freedesktop.org/gstreamer/gst-plugins-good/-/blob/master/ext/pulse/pulsesink.c#L2410) is called. In this method [`gst_pulse_cvolume_from_linear()`
](https://gitlab.freedesktop.org/gstreamer/gst-plugins-good/-/blob/master/ext/pulse/pulseutil.c#L340) is called.

Now it gets interesting. This function is a one-liner:
```c
void
gst_pulse_cvolume_from_linear (pa_cvolume * v, unsigned channels,
    gdouble volume)
{
  pa_cvolume_set (v, channels, pa_sw_volume_from_linear (volume));
}
```

When we call `pa_sw_volume_from_linear (0.05)`, which returns a `pa_volume_t` the result is `0.36840`!
This corresponds with the 5% volume in exaile and 37% volume in pavucontrol.

The type `pa_volume_t` is an alias for `uint32_t`. And it happens to be that when interpreted as `double`, the value is `0.36840`. So, somewhere in pulseaudio gstreamer plugin, some type conversion is going wrong, causing this bug as we see in exaile.

The patch above still solves the issue. However, it should actually check if the sink is pulseaudio, and only then do the volume scale conversion.

**Comment by sjohannes** [comment]
I think this is not a bug in GStreamer's pulseaudio plugin; it's just that GStreamer and PulseAudio use different volume scales. For example, these two commands play audio at the same level:

    gst-launch-1.0 playbin3 volume=0.2 audio-sink=alsasink uri='https://interactive-examples.mdn.mozilla.net/media/cc0-audio/t-rex-roar.mp3'
    gst-launch-1.0 playbin3 volume=0.2 audio-sink=pulsesink uri='https://interactive-examples.mdn.mozilla.net/media/cc0-audio/t-rex-roar.mp3'

Firefox also uses the same scale as GStreamer (even though it doesn't use GStreamer anymore). You can test this by playing a file on Exaile and on Firefox using the same volume percentage, they should play at the same audio level.

**Comment by unkie** [comment]
The Gstreamer documentation on [GStreamVolume](https://gstreamer.freedesktop.org/documentation/audio/gststreamvolume.html?gi-language=c) says "This volume is always a linear factor". This would imply that the pulseaudio plugin would have to take care of scale conversion.
